test/test_distribution_generic.py

```diff
@@ -0,0 +1,433 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+from collections import OrderedDict
+from importlib import import_module
+
+import numpy as np
+import pytest
+
+import funsor
+import funsor.ops as ops
+from funsor.distribution import BACKEND_TO_DISTRIBUTIONS_BACKEND
+from funsor.integrate import Integrate
+from funsor.interpreter import interpretation
+from funsor.terms import Variable, lazy, to_data, to_funsor
+from funsor.testing import assert_close, check_funsor, rand, randint, randn, random_scale_tril, xfail_if_not_implemented  # noqa: F401,E501
+from funsor.util import get_backend
+
+
+_ENABLE_MC_DIST_TESTS = int(os.environ.get("FUNSOR_ENABLE_MC_DIST_TESTS", 0))
+
+pytestmark = pytest.mark.skipif(get_backend() == "numpy",
+                                reason="numpy does not have distributions backend")
+if get_backend() != "numpy":
+    dist = import_module(BACKEND_TO_DISTRIBUTIONS_BACKEND[get_backend()])
+    backend_dist = dist.dist
+
+    class _fakes:
+        """alias for accessing nonreparameterized distributions"""
+        def __getattribute__(self, attr):
+            if get_backend() == "torch":
+                return getattr(backend_dist.testing.fakes, attr)
+            elif get_backend() == "jax":
+                return getattr(dist, "_NumPyroWrapper_" + attr)
+            raise ValueError(attr)
+
+    FAKES = _fakes()
+
+
+##################################################
+# Test cases
+##################################################
+
+class DistTestCase:
+
+    def __init__(self, raw_dist, raw_params, expected_value_domain):
+        self.raw_dist = raw_dist
+        self.raw_params = raw_params
+        self.expected_value_domain = expected_value_domain
+        for name, raw_param in self.raw_params:
+            if get_backend() != "numpy":
+                # we need direct access to these tensors for gradient tests
+                setattr(self, name, eval(raw_param))
+
+    def __str__(self):
+        return self.raw_dist + " " + str(self.raw_params)
+
+    def __hash__(self):
+        return hash((self.raw_dist, self.raw_params, self.expected_value_domain))
+
+
+TEST_CASES = []
+
+for batch_shape in [(), (5,), (2, 3)]:
+
+    # BernoulliLogits
+    TEST_CASES += [DistTestCase(
+        "backend_dist.Bernoulli(logits=case.logits)",
```

Comment:
Note I've chosen to encode test cases with strings of Python code. This should allow us to apply these generic tests even to quite complicated backend distribution expressions, e.g. `TransformedDistribution(Normal(loc, scale).to_event(1), [TanhTransform,]).mask(mask)`